src/main/java/graphql/execution/AsynchronousExecutionStrategy.java

```diff
@@ -0,0 +1,223 @@
+
+        DataFetchingEnvironment environment = new DataFetchingEnvironmentImpl(
+                        parameters.source(),
+                        argumentValues,
+                        executionContext.getRoot(),
+                        fields,
+                        fieldType,
+                        type,
+                        executionContext.getGraphQLSchema(),
+                        executionContext.getFragmentsByName(),
+                        executionContext.getExecutionId(),
+                        fieldCollector);
+
+        Instrumentation instrumentation = executionContext.getInstrumentation();
+
+        InstrumentationContext<ExecutionResult>
+                        fieldCtx = instrumentation.beginField(new FieldParameters(executionContext, fieldDef, environment));
+
+        InstrumentationContext<Object> fetchCtx = instrumentation.beginFieldFetch(new FieldFetchParameters(executionContext, fieldDef, environment));
+        Object resolvedValue = null;
+
+        CompletableFuture<Object> dataFetcherResult = null;
+        try {
+            resolvedValue = fieldDef.getDataFetcher().get(environment);
+
+            if(resolvedValue instanceof CompletionStage) {
+                dataFetcherResult = (CompletableFuture) resolvedValue;
+            } else {
+                dataFetcherResult = CompletableFuture.completedFuture(resolvedValue);
+            }
+        } catch (Exception e) {
+            log.warn("Exception while fetching data", e);
+            dataFetcherResult = new CompletableFuture();
+            dataFetcherResult.completeExceptionally(e);
+        }
+
+        return dataFetcherResult.handleAsync((value,th)-> {
+            if(th != null) {
+                log.warn("Exception while fetching data", th);
+                handleDataFetchingException(executionContext, fieldDef, argumentValues, new ExecutionException(th));
+                fetchCtx.onEnd(th);
+            }
+
+            TypeInfo fieldTypeInfo = newTypeInfo()
+                            .type(fieldType)
+                            .parentInfo(parameters.typeInfo())
+                            .build();
+
+            ExecutionParameters newParameters = ExecutionParameters.newParameters()
+                            .typeInfo(fieldTypeInfo)
+                            .fields(parameters.fields())
+                            .arguments(argumentValues)
+                            .source(value).build();
+
+            return newParameters;
+        }).thenComposeAsync(newParameters -> completeValueAsync(executionContext, newParameters,
+                        fields));
+
+
+    }
+    protected CompletionStage<ExecutionResult> completeValueAsync(ExecutionContext executionContext, ExecutionParameters parameters, List<Field> fields) {
+        TypeInfo typeInfo = parameters.typeInfo();
+        Object result = parameters.source();
+        GraphQLType fieldType = parameters.typeInfo().type();
+
+        if (result == null) {
```

Comment:
fyi - there are a few PRS in the pipe that will make this code break.  But thats often always the case